### PR TITLE
Recursively fetch the thread for receipts & notifications.

### PR DESCRIPTION
**Pull request by clokep**
This is a late-ish addition to MSC3771 & MSC3773 that wasn't in the original pull requests (#13776 and #13782).

The idea is that an event is only part of a thread if you can follow a chain of relations up to a thread root.

Depends on #13776 which is approved, but unmerged due to version bumps.

Part of #12550.

### CODE REVIEW

File: synapse/push/bulk_push_rule_evaluator.py [hunk 1]
```diff
@@ -283,8 +283,13 @@ async def action_for_event_by_user(
                 relation.parent_id,
                 itertools.chain(*(r.rules() for r in rules_by_user.values())),
             )
+            # Recursively attempt to find the thread this event relates to.
             if relation.rel_type == RelationTypes.THREAD:
                 thread_id = relation.parent_id
+            else:
+                # Since the event has not yet been persisted we check whether
+                # the parent is parent of a thread.
```
Comment (erikjohnston) [inline synapse/push/bulk_push_rule_evaluator.py:291]:
```suggestion
                # the parent is child of a thread.
```
?

File: synapse/storage/databases/main/relations.py [hunk 2]
```diff
@@ -832,6 +832,40 @@ def _get_event_relations(
...
+        """
+        sql = """
+            WITH RECURSIVE related_events AS (
+                SELECT event_id, relates_to_id, relation_type
+                FROM event_relations
+                WHERE event_id = ?
+                UNION SELECT e.event_id, e.relates_to_id, e.relation_type
+                FROM event_relations e
+                INNER JOIN related_events r ON r.relates_to_id = e.event_id
+            ) SELECT relates_to_id FROM related_events WHERE relation_type = 'm.thread';
```
Comment (erikjohnston) [inline synapse/storage/databases/main/relations.py:856]: Can we add a comment that event relations form a tree, and so you should only ever get 0 or 1 results from this?

**Review by erikjohnston — COMMENTED** [review]
Otherwise I think it looks good

File: synapse/push/bulk_push_rule_evaluator.py (hunk 1 above)
Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py:291]: I think it is meant to "we check whether the parent is *part* of a thread."

File: synapse/storage/databases/main/relations.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/relations.py:858]: I added a comment above the query. There's also a `TODO` below about ensuring only a single result. Not sure if we care to do anything about that?

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**